osquery/tables/yara/yara.cpp

```diff
@@ -39,71 +40,221 @@ FLAG(bool,
+  auto parser = Config::getParser("yara");
+  if (isNull(parser)) {
+    LOG(ERROR) << "YARA config parser plugin has no pointer";
+    return nullptr;
+  }
+
+  std::shared_ptr<YARAConfigParserPlugin> yaraParser = nullptr;
+  try {
+    yaraParser = std::dynamic_pointer_cast<YARAConfigParserPlugin>(parser);
+  } catch (const std::bad_cast& e) {
+    LOG(ERROR) << "Error casting yara config parser plugin";
+    return nullptr;
+  }
+
+  return yaraParser;
+}
+
 void doYARAScan(YR_RULES* rules,
                 const std::string& path,
                 QueryData& results,
-                const std::string& group,
+                YaraScanType yc,
                 const std::string& sigfile) {
-  Row r;
+  Row row;
 
   // These are default values, to be updated in YARACallback.
-  r["count"] = INTEGER(0);
-  r["matches"] = std::string("");
-  r["strings"] = std::string("");
-  r["tags"] = std::string("");
+  row["count"] = INTEGER(0);
+  row["matches"] = SQL_TEXT("");
+  row["strings"] = SQL_TEXT("");
+  row["tags"] = SQL_TEXT("");
+  row["sig_group"] = SQL_TEXT("");
+  row["sigfile"] = SQL_TEXT("");
+  row["sigrule"] = SQL_TEXT("");
 
   // This could use target_path instead to be consistent with yara_events.
-  r["path"] = path;
-  r["sig_group"] = std::string(group);
-  r["sigfile"] = std::string(sigfile);
+  row["path"] = path;
+
+  switch (yc) {
+  case YC_GROUP:
+    row["sig_group"] = SQL_TEXT(sigfile);
+    break;
+  case YC_FILE:
+    row["sigfile"] = SQL_TEXT(sigfile);
+    break;
+  case YC_RULE:
+    row["sigrule"] = SQL_TEXT(sigfile);
+    break;
+  case YC_NONE:
+    break;
+  }
 
   // Perform the scan, using the static YARA subscriber callback.
   int result = yr_rules_scan_file(
-      rules, path.c_str(), SCAN_FLAGS_FAST_MODE, YARACallback, (void*)&r, 0);
+      rules, path.c_str(), SCAN_FLAGS_FAST_MODE, YARACallback, (void*)&row, 0);
   if (result == ERROR_SUCCESS) {
-    results.push_back(std::move(r));
+    results.push_back(std::move(row));
+  }
+}
+
+Status genYaraRuleFromFile(QueryContext& queryContext,
+                           YaraScanContext& scanContext) {
+  auto yaraParser = getYaraParser();
+  if (isNull(yaraParser)) {
+    return Status::failure("YARA config parser plugin has no pointer");
+  }
+
+  auto& rules = yaraParser->rules();
+  auto sigfiles = queryContext.constraints["sigfile"].getAll(EQUALS);
+
+  for (const auto& file : sigfiles) {
+    // Check if the signature file has been used/ compiled
+    if (rules.count(hashStr(file, YC_FILE)) > 0) {
+      scanContext.insert(std::make_pair(YC_FILE, file));
+      continue;
+    }
+
+    // If this is a relative path append the default yara search path.
+    auto path = (file[0] != '/') ? (kYARAHome + file) : file;
+
+    YR_RULES* tmp_rules = nullptr;
+    auto status = compileSingleFile(path, &tmp_rules);
+    if (!status.ok()) {
+      LOG(WARNING) << "YARA compile error : " << status.toString();
```

Comment:
Prefer consistency `"YARA compile error: " << status.toString();` we only need a space after the semicolon.